### ISSUE TITLE
How to prefer type `Object<>`, but otherwise `object`

### ISSUE THREAD

**Issue by jaydenseric** [gajus/eslint-plugin-jsdoc#709]
[TypeScript doesn't recognise `object<>`, but understands `Object<>`](https://github.com/microsoft/TypeScript/issues/20555). By default, `eslint-plugin-jsdoc` prefers the `object` type over `Object`. My intent is to prefer `object` when the type is not using `<>`, but if it is, prefer `Object<>`.

1. Is TypeScript / VS Code intellisense behavior incorrect, and we should push them for a speedy fix?
2. Should the defaults take this TypeScript / VS Code intellisense behavior into account, and prefer `Object<>`?
3. How can that manually be configured, today?

## Expected behavior

By default, the types enforced with `jsdoc/check-types` should be compatible with TypeScript / VS Code intellisense.

## Actual behavior

The `object<>` type is not.

## ESLint Config

```js
{
  plugins: ['jsdoc'],
  rules: {
    'jsdoc/check-types': 'error'
  }
}
```

## ESLint sample

To see the problem with the default prefered types, this will be an error:

```js
/** @typedef {Object<string, boolean>} foo */
```

<img width="630" alt="Screen Shot 2021-03-30 at 12 43 55 pm" src="https://user-images.githubusercontent.com/1754873/112920948-aae03480-9155-11eb-9886-62e2665fe88e.png">

To try to allow `Object<>`:

```js
{
  plugins: ['jsdoc'],
  settings: {
    jsdoc: {
      preferredTypes: {
        'object<>': 'Object',
      },
    },
  },
  rules: {
    'jsdoc/check-types': 'error'
  }
}
```

Unfortunately, that results a never ending circle of autofix causing another error:

<img width="632" alt="Screen Shot 2021-03-30 at 12 40 42 pm" src="https://user-images.githubusercontent.com/1754873/112920703-360cfa80-9155-11eb-8535-9f53d9944ba1.png">

<img width="631" alt="Screen Shot 2021-03-30 at 12 40 23 pm" src="https://user-images.githubusercontent.com/1754873/112920709-37d6be00-9155-11eb-8c71-efe1ba5ccdd9.png">

## Environment

- Node version: v15.12.0
- ESLint version v7.23.0
- `eslint-plugin-jsdoc` version: v32.3.0

**Comment by brettz9** [comment]
> [TypeScript doesn't recognise `object<>`, but understands `Object<>`](https://github.com/microsoft/TypeScript/issues/20555). By default, `eslint-plugin-jsdoc` prefers the `object` type over `Object`. My intent is to prefer `object` when the type is not using `<>`, but if it is, prefer `Object<>`.
> 
>     1. Is TypeScript / VS Code intellisense behavior incorrect, and we should push them for a speedy fix?

No, it's ok. We just felt that requiring "object" to include `Object.create(null)` was a better match since the lower-case form didn't suggest it needed to descend from `Object` (similar to how using `Boolean` and `Number` suggests wrapped objects).

`object` is for any non-primitive type per https://stackoverflow.com/a/49465172/271577 , but it seems it is not a generic (though neither is `Object`, requiring `Record` for some reason--I'm not really deep into TS, but I'd think using the doc examples in the playground would show the jsdoc type on hover instead of "any".).

But I'd generally be inclined to presume something like that, esp. being out for a while, is working as intended.

>     2. Should the defaults take this TypeScript / VS Code intellisense behavior into account, and prefer `Object<>`?

Yes, that is a good idea, though it seems it should perhaps be `Object.<>` with a dot since it is shown that way at https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#type (it mentions standard JSDoc but JSDoc doesn't really advertise support without the dot). We can look to the mode/parser if we get the underlying support.

>     3. How can that manually be configured, today?

Unfortunately, we can't do this so easily with jsdoctypeparser's current AST traverser. It does not pass along info on the children or their being children (only the parent), so we'd have to do our own traversal.

**Comment by brettz9** [comment]
It seems we already make an exception to allow `Object` when the `mode` is "typescript" (or the parser is `@typescript/eslint-parser`).

**Comment by brettz9** [comment]
@jaydenseric : Maybe you want to offer your opinion on #800 ?

**Comment by brettz9** [comment]
Closing in favor of carrying forward discussion on #800 . Hopefully to have a fix soon.
